OrbitGl/ProcessData.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ProcessData.h"
+
+#include <memory>
+#include <vector>
+
+#include "OrbitBase/Result.h"
+#include "absl/strings/str_format.h"
+#include "process.pb.h"
+
+using orbit_client_protos::FunctionInfo;
+using orbit_grpc_protos::ModuleSymbols;
+using orbit_grpc_protos::ProcessInfo;
+
+void ProcessData::UpdateModuleInfos(
+    const std::vector<orbit_grpc_protos::ModuleInfo>& module_infos) {
+  module_memory_map_.clear();
+  start_addresses_.clear();
+
+  for (const auto& module_info : module_infos) {
+    {
+      const auto [it, success] = module_memory_map_.try_emplace(
+          module_info.file_path(),
+          std::make_unique<MemorySpace>(module_info.address_start(), module_info.address_end()));
+      CHECK(success);
+    }
+    {
+      const auto [it, success] =
+          start_addresses_.try_emplace(module_info.address_start(), module_info.file_path());
+      CHECK(success);
+    }
+  }
+}
+
+ErrorMessageOr<std::pair<std::string, uint64_t>> ProcessData::FindModuleByAddress(
+    uint64_t absolute_address) const {
+  if (start_addresses_.empty()) {
+    return ErrorMessage(absl::StrFormat("Unable to find module for address %016" PRIx64
+                                        ": No modules loaded by process %s",
+                                        absolute_address, name()));
+  }
+
+  ErrorMessage not_found_error =
+      ErrorMessage(absl::StrFormat("Unable to find module for address %016" PRIx64
+                                   ": No module loaded at this address by process %s",
+                                   absolute_address, name()));
+
+  auto it = start_addresses_.upper_bound(absolute_address);
+  if (it == start_addresses_.begin()) return not_found_error;
+
+  --it;
+  const std::string& module_path = it->second;
+  const MemorySpace* memory_space = module_memory_map_.at(module_path).get();
+  CHECK(absolute_address >= memory_space->start);
+  if (absolute_address > memory_space->end) return not_found_error;
+
+  return std::make_pair(module_path, memory_space->start);
+}
+
+void ProcessData::LoadSymbols(ModuleData* module, const ModuleSymbols& module_symbols) const {
+  uint64_t module_start_address = module_memory_map_.at(module->file_path())->start;
+  module->LoadSymbols(module_symbols, module_start_address);
+}
+
+std::unique_ptr<ProcessData> ProcessData::CreateCopy() const {
```

Comment:
What is the use-case here? Shouldn't we aim for just making ProcessData copyable?

Reply:
I do not want to make ProcessData trivially copyable, because we are using ProcessData* in a lot of places and have to be careful to not accidentally create a copy. That is why I would like the copy constructor to stay deleted. I want to force us to jump through the hoop of calling this function in the rare case that we actually need to make a copy (currently only for a capture thread local instance of ProcessData). 

This is a personal design choice and I am unsure if my reasoning makes sense or if I am creating unnecessary complexity. Any opinions are welcome :)

Reply:
Fair enough for me. Thanks for explaining.